runtime/sema/check_expression.go

```diff
@@ -256,45 +256,98 @@ func (checker *Checker) visitIndexExpression(
 		return InvalidType
 	}
 
-	// Check if the type instance is actually indexable. For most types (e.g. arrays and dictionaries)
-	// this is known statically (in the sense of this host language (Go), not the implemented language),
-	// i.e. a Go type switch would be sufficient.
-	// However, for some types (e.g. reference types) this depends on what type is referenced
-
-	indexedType, ok := targetType.(ValueIndexableType)
-	if !ok || !indexedType.isValueIndexableType() {
+	reportNonIndexable := func(t Type) {
 		checker.report(
 			&NotIndexableTypeError{
-				Type:  targetType,
+				Type:  t,
 				Range: ast.NewRangeFromPositioned(checker.memoryGauge, targetExpression),
 			},
 		)
-
-		return InvalidType
 	}
 
-	indexingType := checker.VisitExpression(
-		indexExpression.IndexingExpression,
-		indexedType.IndexingType(),
-	)
-
-	if isAssignment && !indexedType.AllowsValueIndexingAssignment() {
-		checker.report(
-			&NotIndexingAssignableTypeError{
-				Type:  indexedType,
-				Range: ast.NewRangeFromPositioned(checker.memoryGauge, targetExpression),
-			},
+	// Check if the type instance is actually indexable. For most types (e.g. arrays and dictionaries)
+	// this is known statically (in the sense of this host language (Go), not the implemented language),
+	// i.e. a Go type switch would be sufficient.
+	// However, for some types (e.g. reference types) this depends on what type is referenced
+	// we cannot use a switch here because in the reference type case we need to be able to "fallthrough",
+	// and Go does not allow fallthroughs in typed switches
+	valueIndexedType, isValueIndexableType := targetType.(ValueIndexableType)
+	typeIndexedType, isTypeIndexableType := targetType.(TypeIndexableType)
```

Comment:
you should be able to move this cast to line 301, saving the overhead if `targetType` is value indexable